spec/datadog/appsec/waf/handle_builder_spec.rb

```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal:true
+
+require "spec_helper"
+
+RSpec.describe Datadog::AppSec::WAF::HandleBuilder do
+  subject(:builder) { described_class.new }
+
+  let(:valid_config) do
+    JSON.parse(File.read("spec/fixtures/valid_config.json"))
+  end
+
+  describe "#build_handle" do
+    context "when at least one valid rule has been loaded" do
+      before do
+        builder.add_or_update_config(config: valid_config, path: "some/path")
+      end
+
+      after do
+        builder.remove_config(path: "some/path")
+      end
+
+      it "returns a Datadog::AppSec::WAF::Handle instance" do
+        expect(builder.build_handle).to be_a(Datadog::AppSec::WAF::Handle)
+      end
+    end
+
+    it "raises LibDDWAF::Error when no valid rule has been loaded" do
+      expect { builder.build_handle }.to raise_error(Datadog::AppSec::WAF::LibDDWAFError, /Could not create handle/)
+    end
+
+    it "raises LibDDWAF::Error when builder has been finalized" do
+      builder.finalize!
+
+      expect do
+        builder.build_handle
+      end.to raise_error(Datadog::AppSec::WAF::HandleBuilderFinalizedError, /Cannot use WAF handle builder after it has been finalized/)
+    end
```

Comment:
Could we move those up, it's hard to read nesting like this because we have context definition above and nesting makes it hard to understand